### PR TITLE
WFLY-9877 Auto-update obsolete protocols to their current variants.

### PR DESCRIPTION
**Pull request by pferraro**
Fixes JGroups configuration compatibility using obsolete protocols.

https://issues.jboss.org/browse/WFLY-9877

### CODE REVIEW

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/logging/JGroupsLogger.java [hunk 1]
```diff
@@ -155,6 +154,6 @@
     IllegalArgumentException failedToResolveSocketBinding(@Cause UnknownHostException cause, OutboundSocketBinding binding);
 
     @LogMessage(level = WARN)
-    @Message(id = 29, value = "Ignoring unrecognized %s properties: %s")
-    void ignoredProperties(String protocol, Map<String, String> properties);
+    @Message(id = 29, value = "Protocol %s is obsolete and will be auto-updated to %s")
+    void legacyProtocol(String legacyProtocol, String targetProtocol);
```
Comment (stuartwdouglas) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/logging/JGroupsLogger.java]: It looks like this message has been in a release, so it should probably be commented out and a new ID assigned to the new message.

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    private static class OperationTransformation implements UnaryOperator<OperationStepHandler> {
+        private final String targetName;
+
+        OperationTransformation(String targetName) {
+            this.targetName = targetName;
+        }
+
+        @Override
+        public OperationStepHandler apply(OperationStepHandler handler) {
+            return (context, operation) -> {
```
Comment (stuartwdouglas) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: This should not be a lambda.

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                Operations.setPathAddress(operation, targetAddress);
+                PathAddress targetRegistrationAddress = address.getParent().append(ProtocolResourceDefinition.WILDCARD_PATH);
+                String operationName = Operations.getName(operation);
+                context.addStep(operation, context.getRootResourceRegistration().getOperationHandler(targetRegistrationAddress, operationName), OperationContext.Stage.MODEL);
+            };
+        }
+    }
+
+    LegacyProtocolResourceDefinition(String name, String targetName, JGroupsModel deprecation, Consumer<ResourceDescriptor> descriptorConfigurator, ResourceServiceBuilderFactory<ChannelFactory> parentBuilderFactory) {
+        super(pathElement(name), descriptorConfigurator.andThen(descriptor -> descriptor.setAddOperationTransformation(new OperationTransformation(targetName)).setOperationTransformation(new OperationTransformation(targetName))), null, parentBuilderFactory);
```
Comment (stuartwdouglas) [inline clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: lambda

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java (hunk 2 above)
Comment (pferraro) [reply clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: Is there really anything to be gained by replacing this with an anonymous class?

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java (hunk 3 above)
Comment (pferraro) [reply clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: Is there really anything to be gained by creating an additional class that implements Consumer<ResourceDescriptor>?

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java (hunk 2 above)
Comment (stuartwdouglas) [reply clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: According to Andrew Dinn:

"As Bill says every lambda requires its own anonymous class and that adds
about 700 /used/ bytes to metaspace. Unfortunately, the cost is more
than that. The JVM currently places Lambda classes in their own class
loader. That is done so they can be garbage collected should the method
handle at the lambda call site be updated. That means the 700 used bytes
get allocated out of a 1.5K memory chunk dedicated to that class loader.
Since the loader has only one class none of the remaining free space can
be used for other metadata. So, the real /used/ count for a lambda is
the full 1.5K."

So basically it saves one class loader and its associated metaspace allocation.

Comment (pferraro) [reply clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: Fixed.

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java (hunk 3 above)
Comment (pferraro) [reply clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/LegacyProtocolResourceDefinition.java]: Fixed.

File: clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/logging/JGroupsLogger.java (hunk 1 above)
Comment (pferraro) [reply clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/logging/JGroupsLogger.java]: Fixed.
